### PR TITLE
Tempo: Metrics summary

### PR DESCRIPTION
**Pull request by joey-grafana**
**What is this feature?**

Adds support for a metrics summary in Tempo.

**Why do we need this feature?**

So users can get a summary of metrics for their traces.

**Who is this feature for?**

Tempo users.

**Special notes for your reviewer:**

This feature exists behind the `metricsSummary` feature toggle.

Screenshot 2023-08-22 at 10 13 51

Please check that:
- [x] It works as expected from a user's perspective.
- [x] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.
@knylander-grafana has offered to updated the docs :)

### CODE REVIEW

File: public/app/plugins/datasource/tempo/datasource.ts
Comment (aocenas) [inline public/app/plugins/datasource/tempo/datasource.ts:216]: I assume this diff is mainly because of reordering the IFs here right? Or did something change apart from the `if (groupBy) {` part?

File: public/app/plugins/datasource/tempo/datasource.ts
Comment (aocenas) [inline public/app/plugins/datasource/tempo/datasource.ts:419]: Why do we start with this? Isn't loading state done automatically as query is started?

File: public/app/plugins/datasource/tempo/datasource.ts
Comment (aocenas) [inline public/app/plugins/datasource/tempo/datasource.ts:424]: Why the `response.state !== LoadingState.Loading`? Is that just bucause we have that `startWith` value?

File: public/app/plugins/datasource/tempo/datasource.ts
Comment (aocenas) [inline public/app/plugins/datasource/tempo/datasource.ts:446]: Do we need to create the data frame if we have error? Don't we handle that reasonably in Explore/Dashboards?

File: public/app/plugins/datasource/tempo/metricsSummary.ts
Comment (aocenas) [inline public/app/plugins/datasource/tempo/metricsSummary.ts:54]: How do we get to this point? It does not seem like Tempo is streaming the response is it?

File: public/app/plugins/datasource/tempo/metricsSummary.ts
Comment (aocenas) [inline public/app/plugins/datasource/tempo/metricsSummary.ts:70]: Please lets not use `map()` to do something else than actually map data. It's very confusing if map is use as an iterator that has side effects. Rather use `for` loop or `forEach()`.

File: public/app/plugins/datasource/tempo/metricsSummary.ts
Comment (aocenas) [inline public/app/plugins/datasource/tempo/metricsSummary.ts:83]: As MutableDataFrame is deprecated we should probably use something else to create the Frames. Either manually or through the new helper functions that exists for it.

File: public/app/plugins/datasource/tempo/metricsSummary.ts
Comment (aocenas) [inline public/app/plugins/datasource/tempo/metricsSummary.ts:113]: Can we move this up top? Seems like we are doing lots of things when we could just return early with this condition.

File: public/app/plugins/datasource/tempo/metricsSummary.ts
Comment (aocenas) [inline public/app/plugins/datasource/tempo/metricsSummary.ts:124]: I guess if we move `if (!data?.length) {` to the top we can then be sure that `frame` always exists if there is `data`?

File: public/app/plugins/datasource/tempo/metricsSummary.ts
Comment (aocenas) [inline public/app/plugins/datasource/tempo/metricsSummary.ts:153]: What does this actually do? I mean where does the `query` show in Grafana when returned like this? First time I am seeing it used so am curious.

File: public/app/plugins/datasource/tempo/SearchTraceQLEditor/GroupByField.tsx
Comment (aocenas) [inline public/app/plugins/datasource/tempo/SearchTraceQLEditor/GroupByField.tsx:72]: I think this is doing a copy 2 times doesn't it? Seems like what we pass to `onChange` should be enough to make this function pure.

File: public/app/plugins/datasource/tempo/SearchTraceQLEditor/GroupByField.tsx
Comment (aocenas) [inline public/app/plugins/datasource/tempo/SearchTraceQLEditor/GroupByField.tsx:80]: Not sure I understand this comment. We do copy of the query and also `replaceAt` does a copy of the array so it does not seem like you are updating this in place

File: public/app/plugins/datasource/tempo/SearchTraceQLEditor/GroupByField.tsx
Comment (aocenas) [inline public/app/plugins/datasource/tempo/SearchTraceQLEditor/GroupByField.tsx:106]: if `id` is a generated UID, I don't think it would be of much help here?

File: public/app/plugins/datasource/tempo/datasource.ts
Comment (joey-grafana) [reply public/app/plugins/datasource/tempo/datasource.ts:216]: Yep this is basically reordering/update because of the group by part.

File: public/app/plugins/datasource/tempo/datasource.ts
Comment (joey-grafana) [reply public/app/plugins/datasource/tempo/datasource.ts:419]: The query is sent off and returns nothing until it has all the data. Perhaps it's not supposed to work like that? This is so the user knows that the request was sent and is in loading state.

File: public/app/plugins/datasource/tempo/datasource.ts
Comment (joey-grafana) [reply public/app/plugins/datasource/tempo/datasource.ts:446]: We do so that we can update the state from loading to no data.

With empty data frame
Screenshot 2023-08-23 at 08 53 38

Without empty data frame
Screenshot 2023-08-23 at 08 53 47

File: public/app/plugins/datasource/tempo/metricsSummary.ts
Comment (joey-grafana) [reply public/app/plugins/datasource/tempo/metricsSummary.ts:54]: We get here via the `startWith` that sets the state to `Loading`. The metrics summary endpoint is not queried via steaming.

File: public/app/plugins/datasource/tempo/metricsSummary.ts
Comment (joey-grafana) [reply public/app/plugins/datasource/tempo/metricsSummary.ts:70]: Ahh yes, good catch.

File: public/app/plugins/datasource/tempo/metricsSummary.ts
Comment (joey-grafana) [reply public/app/plugins/datasource/tempo/metricsSummary.ts:153]: This is the query that is opened in explore when the user clicks on a dynamic attribute cell in the table.

File: public/app/plugins/datasource/tempo/SearchTraceQLEditor/GroupByField.tsx
Comment (joey-grafana) [reply public/app/plugins/datasource/tempo/SearchTraceQLEditor/GroupByField.tsx:80]: Honestly copied this method from `<TraceQLSearch />`. The logic checks out and works in the UI so I've removed the comment.

File: public/app/plugins/datasource/tempo/SearchTraceQLEditor/GroupByField.tsx
Comment (joey-grafana) [reply public/app/plugins/datasource/tempo/SearchTraceQLEditor/GroupByField.tsx:106]: It's to make a distinction between multiple group by filters. Not sure what else to use here that would make it clear that this is a unique filter and different from the other group by filters.

File: public/app/plugins/datasource/tempo/datasource.ts
Comment (aocenas) [reply public/app/plugins/datasource/tempo/datasource.ts:419]: I tried some changes cause I wasn't exactly sure but you can see them here https://github.com/grafana/grafana/pull/73666 and it seems like it works fine. It even get rid a short flicker when it seems we would show `0 series retrieved` which I assume is because of that one loading frame returned. I think loading state for simple requests, which just return single response (so no streaming, multiresponse or pooling) should be handled by explore/dashboard and you don't need to do anything special there.

File: public/app/plugins/datasource/tempo/SearchTraceQLEditor/GroupByField.tsx
Comment (aocenas) [reply public/app/plugins/datasource/tempo/SearchTraceQLEditor/GroupByField.tsx:104]: So I mean this is mainly for screen readers, right? So if this reads `Select scope for dJ2Az1` or something, I don't think that helps the user that much.

File: public/app/plugins/datasource/tempo/metricsSummary.ts
Comment (aocenas) [reply public/app/plugins/datasource/tempo/metricsSummary.ts]: Sorry still don't get it. I mean don't we add data links to the frame to make the click work?

Comment (joey-grafana) [reply public/app/plugins/datasource/tempo/metricsSummary.ts]: This method constructs the query that goes in the link, so IIUC then yes.

Screenshot 2023-08-23 at 10 52 38

Screenshot 2023-08-23 at 10 52 26

File: public/app/plugins/datasource/tempo/SearchTraceQLEditor/GroupByField.tsx
Comment (joey-grafana) [reply public/app/plugins/datasource/tempo/SearchTraceQLEditor/GroupByField.tsx:104]: I agree it would be better to have something that makes more sense for screen readers, but without the id here, every group by filter will just read as `Select scope`. To simplify this and to make sure the screen reader user knows which filter is which, I've just added the filter index here as opposed to its id.

File: public/app/plugins/datasource/tempo/datasource.ts
Comment (joey-grafana) [reply public/app/plugins/datasource/tempo/datasource.ts]: Decided it's prob better to not add custom loading state logic for just this use case to the UI.

File: public/app/plugins/datasource/tempo/metricsSummary.ts
Comment (aocenas) [reply public/app/plugins/datasource/tempo/metricsSummary.ts]: Ah ok sorry got confused, thought this was some separate dataFrame config for some other purpose.

**Review by aocenas — APPROVED** [review]
Approving so it's not blocked while I am away. I think we talked about all the major points so after those are addressed this can be merged

File: public/app/plugins/datasource/tempo/metricsSummary.ts
Comment (joey-grafana) [reply public/app/plugins/datasource/tempo/metricsSummary.ts]: No prob!

**Pull request merged**
